test/hardhat/irm/Irm.spec.ts

```diff
@@ -0,0 +1,95 @@
+import { AbiCoder, keccak256, toBigInt } from "ethers";
+import hre from "hardhat";
+import _range from "lodash/range";
+import { AdaptativeCurveIrm } from "types";
+import { MarketParamsStruct } from "types/lib/morpho-blue/src/interfaces/IIrm";
+
+import { SignerWithAddress } from "@nomicfoundation/hardhat-ethers/signers";
+import { setNextBlockTimestamp } from "@nomicfoundation/hardhat-network-helpers/dist/src/helpers/time";
+
+let seed = 42;
+const random = () => {
+  seed = (seed * 16807) % 2147483647;
+
+  return (seed - 1) / 2147483646;
+};
+
+const identifier = (marketParams: MarketParamsStruct) => {
+  const encodedMarket = AbiCoder.defaultAbiCoder().encode(
+    ["address", "address", "address", "address", "uint256"],
+    Object.values(marketParams),
+  );
+
+  return Buffer.from(keccak256(encodedMarket).slice(2), "hex");
+};
+
+const logProgress = (name: string, i: number, max: number) => {
+  if (i % 10 == 0) console.log("[" + name + "]", Math.floor((100 * i) / max), "%");
+};
+
+const randomForwardTimestamp = async () => {
+  const block = await hre.ethers.provider.getBlock("latest");
+  const elapsed = random() < 1 / 2 ? 0 : (1 + Math.floor(random() * 100)) * 12; // 50% of the time, don't go forward in time.
+
+  const newTimestamp = block!.timestamp + elapsed;
+
+  await setNextBlockTimestamp(block!.timestamp + elapsed);
+
+  return newTimestamp;
+};
+
+describe("irm", () => {
+  let admin: SignerWithAddress;
+
+  let irm: AdaptativeCurveIrm;
+
+  let marketParams: MarketParamsStruct;
+
+  beforeEach(async () => {
+    [admin] = await hre.ethers.getSigners();
+
+    const AdaptativeCurveIrmFactory = await hre.ethers.getContractFactory("AdaptativeCurveIrm", admin);
+
+    irm = await AdaptativeCurveIrmFactory.deploy(
+      await admin.getAddress(),
+      4000000000000000000n,
+      1585489599188n,
+      900000000000000000n,
+      317097919n,
+    );
+
+    const irmAddress = await irm.getAddress();
+
+    marketParams = {
+      // Non-zero address to include calldata gas cost.
+      collateralToken: irmAddress,
+      loanToken: irmAddress,
+      oracle: irmAddress,
+      irm: irmAddress,
+      lltv: 0,
+    };
+
+    hre.tracer.nameTags[irmAddress] = "IRM";
+  });
+
+  it("should simulate gas cost [main]", async () => {
+    for (let i = 0; i < 200; ++i) {
+      logProgress("main", i, 200);
+
+      const lastUpdate = await randomForwardTimestamp();
+
+      const totalSupplyAssets = BigInt.WAD * toBigInt(1 + Math.floor(random() * 100));
+      const totalBorrowAssets = totalSupplyAssets.min(BigInt.WAD * toBigInt(1 + Math.floor(random() * 50)));
```

Comment:
This means that full utilization is tested 25% of the time, and that utilizations from 50% to 100% (not included) are tested less often. One alternative is to simply take `random() * totalSupplyAssets`